File: 0x000a_10.Regular_Expression_Matching/solution.java
class Solution {
    public boolean isMatch(String s, String p) {
        boolean[][] dp = new boolean[s.length()+1][p.length()+1];
        dp[0][0] = true;
        for (int i = 0; i < p.length(); i++) {
            if (i >= 1 && p.charAt(i) == '*') dp[0][i+1] = dp[0][i-1];  
        }
        for (int i = 0; i < s.length(); i++) {
            for (int j = 0; j < p.length(); j++) {
                if (j >= 1 && p.charAt(j) == '*') {
                    dp[i+1][j+1] = dp[i+1][j-1];
                    if (cmp(s.charAt(i), p.charAt(j-1))) dp[i+1][j+1] = dp[i][j-1] || dp[i][j+1] || dp[i+1][j+1];
                } else {
                    dp[i+1][j+1] = (cmp(s.charAt(i), p.charAt(j)) && dp[i][j]);
                }
            }
        }
        return dp[s.length()][p.length()];
    }

    private boolean cmp(char sc, char pc) {
        if (pc == '.') return true;
        return sc == pc;
    }
}

File: 0x000a_10.Regular_Expression_Matching/solution.py
class Solution:
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        ls, lp = len(s), len(p)
        dp = [[False] * (lp + 1) for _ in range(ls + 1)]
        dp[0][0] = True
        for j in range(lp):
            if p[j] == "*":
                dp[0][j + 1] = dp[0][j - 1]
        for i in range(ls):
            for j in range(lp):
                if p[j] == "*":
                    dp[i + 1][j + 1] = dp[i + 1][j - 1] or (
                        dp[i][j + 1] and (p[j - 1] == "." or p[j - 1] == s[i])
                    )
                elif p[j] == ".":
                    dp[i + 1][j + 1] = dp[i][j]
                else:
                    dp[i + 1][j + 1] = dp[i][j] and s[i] == p[j]
        return dp[-1][-1]
